paasta_tools/secret_tools.py

```diff
@@ -103,3 +104,72 @@ def get_secret_hashes(
                 secret_environment=secret_environment,
             )
     return secret_hashes
+
+
+def decrypt_secret_environment_for_service(
+    secret_env_vars: Dict[str, str],
+    service_name: str,
+    secret_provider_name: str,
+    soa_dir: str,
+    cluster_name: str,
+    secret_provider_kwargs: Dict[str, Any],
+) -> Dict[str, str]:
+    if not secret_env_vars:
+        return {}
+
+    secret_provider = get_secret_provider(
+        secret_provider_name=secret_provider_name,
+        soa_dir=soa_dir,
+        service_name=service_name,
+        cluster_names=[cluster_name],
+        secret_provider_kwargs=secret_provider_kwargs,
+    )
+    return secret_provider.decrypt_environment(secret_env_vars)
+
+
+def decrypt_secret_environment_variables(
+    secret_provider_name: str,
+    environment: Dict[str, str],
+    soa_dir: str,
+    service_name: str,
+    cluster_name: str,
+    secret_provider_kwargs: Dict[str, Any],
+) -> Dict[str, str]:
+    decrypted_secrets = {}
+    service_secret_env = {}
+    shared_secret_env = {}
+    for k, v in environment.items():
+        if is_secret_ref(v):
+            if is_shared_secret(v):
+                shared_secret_env[k] = v
+            else:
+                service_secret_env[k] = v
+    secret_provider_kwargs["vault_num_uses"] = len(service_secret_env) + len(
+        shared_secret_env
+    )
+
+    try:
+        decrypted_secrets.update(
+            decrypt_secret_environment_for_service(
+                service_secret_env,
+                service_name,
+                secret_provider_name,
+                soa_dir,
+                cluster_name,
+                secret_provider_kwargs,
+            )
+        )
+        decrypted_secrets.update(
+            decrypt_secret_environment_for_service(
```

Comment:
i know you're preserving original functionality, but wouldn't this call override secrets from previous `.update`? seems like the order should be switched, more specific secrets written last, overriding more general secrets.

Reply:
Hm, I wrote the original functionality, and didn't think about this...  In practice I don't think these dicts will share any keys, because the keys come from the env dict in yelpsoa-configs, which will already be de-deduplicated. For example, if someone puts
```
env:
    SECRET1: SECRET(first)
    SECRET1: SHARED_SECRET(second)
```
the service_secret_env at this point will have one or the other value for SECRET1, depending on how the yaml parsing works.

I think that means the order of updates doesn't matter here, but switching them is fine if that makes it more readable.

Reply:
good point, i missed that `service_secret_env` and `shared_secret_env` are built from the same dictionary so their keys can't overlap, resolving this.